Use RuntimeHelpers.CreateSpan for array initialization with relevant primitives

stephentoub opened the pull request:
The compiler currently optimizes array creation where:
- The array is immediately stored as a `ReadOnlySpan<T>` such that it's not exposed in any writeable manner
- The array is initialized entirely with constant values
- The element type is a single byte primitive (byte, bool, sbyte)

In such a situation, it creates a PrivateImplementationDetails field containing the data, and replaces the array creation with a span that points directly to the data on PrivateImplementationDetails.

This PR extends that optimization to multi-byte primitives: short, ushort, char, int, uint, long, ulong, float, and double.  If the target platform has the RuntimeHelpers.CreateSpan method, it is used to handle endianness fixups.  If the target platform lacks that method, the compiler emits code to cache the array in PrivateImplementationDetails and wrap a span around that lazily-initialized array.

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -370,36 +370,59 @@ private static bool IsMultidimensionalInitializer(ImmutableArray<BoundExpression
+                        // For primitives that are larger than a single byte, endianness needs to be considered,
+                        // which requires then using the runtime's RuntimeHelpers.CreateSpan if it's available.
+                        createSpan = (MethodSymbol)_module.Compilation.GetWellKnownTypeMember(WellKnownMember.System_Runtime_CompilerServices_RuntimeHelpers__CreateSpanRuntimeFieldHandle);
+                        if (createSpan is null)
```
>createSpan is null [](http://example.com/codeflow?start=28&length=18)

In general, not null value doesn't mean the member is good to use. We should check if there are still errors associated with it. For example, check ```GetWellKnownTypeMember``` helpers in the Binder.

stephentoub replied:
There's existing code a few lines before this that's doing:
```C#
var ctor = ((MethodSymbol)this._module.Compilation.GetWellKnownTypeMember(WellKnownMember.System_ReadOnlySpan_T__ctor_Pointer));
```
Is that also wrong?

AlekseyTs replied:
> There's existing code a few lines before this that's doing:
> ...
> without checking diagnostics. Is that also wrong? 

That code also performs an incomplete check for a usage of a well known member.

>  I was just copying what's already being done in this file.

 I understand that, perhaps, the likelihood to find a method that is still "bad" is extremely low for these helpers, but it is better to always perform full validation. One of the reasons is the fact that code is often copied for other helpers and between compiler layers.

jcouv commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -370,36 +370,59 @@ private static bool IsMultidimensionalInitializer(ImmutableArray<BoundExpression
         }
 
-        private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExpression wrappedExpression, bool used, bool inPlace)
+        private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExpression wrappedExpression, bool used, BoundExpression inPlaceTarget, out bool avoidInPlace)
```
FYI, conflicts incoming: [https://github.com/dotnet/roslyn/pull/61532](https://github.com/dotnet/roslyn/pull/61532)

stephentoub replied:
Resolved.

stephentoub commented:
> including merge commits

Most of the code this touches had to be rewritten in the face of the conflicting commit that got merged.  I don't believe the effort required to review with a rebase or a merge would have been significantly different.  But I will try to refrain from rebases in the future.

MichalPetryka commented:
Would be nice if `ReadOnlySpan<char> span = "text"` would also be optimized, since it's a much more convenient syntax for strings and would be analogous to the utf8 string feature.

stephentoub commented:
> Would be nice if ReadOnlySpan<char> span = "text"; would also be optimized, since it's a much more convenient syntax for strings and would be analogous to the utf8 string feature.

What is it you'd hope it would do it's not already?
[SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHunGAlGNgAmAeQB2AGwCeAZQAO2MQB4wAC2xQAfAwCy5ABQBKBgF5tAIgwwEGcwG4aAXyA==)

Neme12 commented:
> Would be nice if ReadOnlySpan<char> span = "text";

Isn't that covered by this PR? It seems that this does check for `char` as well.

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -372,37 +372,61 @@ private static bool IsMultidimensionalInitializer(ImmutableArray<BoundExpression
 
             var ctor = ((MethodSymbol?)this._module.Compilation.GetWellKnownTypeMember(WellKnownMember.System_ReadOnlySpan_T__ctor_Pointer));
-            if (ctor == null)
+            if (ctor == null || ctor.GetUseSiteInfo().DiagnosticInfo?.Severity == DiagnosticSeverity.Error)
```
>|| ctor.GetUseSiteInfo().DiagnosticInfo?.Severity == DiagnosticSeverity.Error [](http://example.com/codeflow?start=29&length=77)

Instead of performing a manual check like this, let's use an existing helper to get the symbol, which is also going to take care of proper dependency tracking:
```
var ctor = (MethodSymbol?)GetWellKnownTypeMember(this._module.Compilation, WellKnownMember.System_ReadOnlySpan_T__ctor_Pointer, _diagnostics, syntax: wrappedExpression.Syntaxl, isOptional: true);
if (ctor == null)
{
    return false;
}
```

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -372,37 +372,61 @@ private static bool IsMultidimensionalInitializer(ImmutableArray<BoundExpression
+                        // For primitives that are larger than a single byte, endianness needs to be considered,
+                        // which requires then using the runtime's RuntimeHelpers.CreateSpan if it's available.
+                        createSpan = (MethodSymbol?)_module.Compilation.GetWellKnownTypeMember(WellKnownMember.System_Runtime_CompilerServices_RuntimeHelpers__CreateSpanRuntimeFieldHandle);
+                        if (createSpan is null || createSpan.GetUseSiteInfo().DiagnosticInfo?.Severity == DiagnosticSeverity.Error)
```
>createSpan.GetUseSiteInfo().DiagnosticInfo?.Severity == DiagnosticSeverity.Error [](http://example.com/codeflow?start=50&length=80)

Similar suggestion to use Binder.GetWellKnownTypeMember

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -446,18 +470,23 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
                     EmitSymbolToken(spanType, wrappedExpression.Syntax);
+                    if (used)
+                    {
+                        EmitExpression(inPlaceTarget, used: true);
```
>EmitExpression(inPlaceTarget, used: true); [](http://example.com/codeflow?start=24&length=42)

It is not obvious what is going on here. Are we potentially going to evaluate an arbitrary complex expression twice?

AlekseyTs replied:
I see that ```inPlaceTarget``` is going to be either a local or parameter. Consider adding ```Debug.Assert(inPlaceTarget is null || TargetIsNotOnHeap(inPlaceTarget), "in-place construction target should not be on heap");``` at the beginning of the function.

stephentoub replied:
Done

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -372,37 +372,61 @@ private static bool IsMultidimensionalInitializer(ImmutableArray<BoundExpression
 #nullable enable
 
-        private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExpression wrappedExpression, bool used, bool inPlace, BoundExpression? start = null, BoundExpression? length = null)
+        private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExpression wrappedExpression, bool used, BoundExpression inPlaceTarget, out bool avoidInPlace, BoundExpression? start = null, BoundExpression? length = null)
```
>BoundExpression inPlaceTarget, out bool avoidInPlace [](http://example.com/codeflow?start=126&length=52)

I think we can benefit from detailed documentation for these parameters. How they are supposed to be used, etc.

stephentoub replied:
Done

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -471,21 +500,50 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+                    _builder.EmitOpCode(ILOpCode.Ldtoken);
+                    _builder.EmitToken(field, wrappedExpression.Syntax, _diagnostics);
+                    _builder.EmitOpCode(ILOpCode.Call, stackAdjustment: 0);
+                    EmitSymbolToken(createSpan, wrappedExpression.Syntax, optArgList: null);
```
>EmitSymbolToken(createSpan, wrappedExpression.Syntax, optArgList: null); [](http://example.com/codeflow?start=20&length=72)

It looks like this code path doesn't handle situation when ```lengthForConstructor``` != ```elementCount```.

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -471,21 +500,50 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+                        _builder.EmitOpCode(ILOpCode.Call, stackAdjustment: -3);
+                        if (used)
+                        {
+                            EmitExpression(inPlaceTarget, used: true);
```
>EmitExpression(inPlaceTarget, used: true); [](http://example.com/codeflow?start=28&length=42)

Similar question about double evaluation.

AlekseyTs commented:
Done with review pass (commit 2), new tests are not reviewed.

stephentoub commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -372,37 +372,61 @@ private static bool IsMultidimensionalInitializer(ImmutableArray<BoundExpression
                 return false;
             }
 
             var ctor = ((MethodSymbol?)this._module.Compilation.GetWellKnownTypeMember(WellKnownMember.System_ReadOnlySpan_T__ctor_Pointer));
```
Done

stephentoub commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -467,17 +488,21 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
             {
-                // emitting a value that no one will see
+                // The caller has specified that we're creating a ReadOnlySpan expression that won't be used.
+                // We needn't emit anything.
```
I don't know. I've kept it where it was before.

Neme12 commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -372,9 +372,23 @@ private static bool IsMultidimensionalInitializer(ImmutableArray<BoundExpression
+        /// <param name="wrappedExpression">The expression being wrapped in a span.</param>
+        /// <param name="used">true if the result of the expression is used; false if it's required only for its side effects.</param>
+        /// <param name="inPlaceTarget">A non-null expression if the construction is initializing an existing local in-place; otherwise, null.</param>
+        /// <param name="avoidInPlace">An output Boolean indicating whether a caller trying to perform in-place initialization should instead prefer to assign the local to a new value.</param>
```
What does in-place initialization mean exactly?

stephentoub replied:
Assigning to an existing local.

stephentoub commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -497,14 +522,27 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+                        // rather than `new ReadOnlySpan<char>(new[] { 'b', 'c' })`.  If such a pattern is found to be
+                        // common, this could be augmented to accomodate it.  For now, we just return false to fail
+                        // to optimize this case.
+                        return false;
```
> the blob will just contain the subarray?

Without additional work, the blob will contain everything and either that would need to be changed to just be the subset or the resulting span would need to be sliced. "Why not support this case". is I don't think such a construction will be at all common.

stephentoub replied:
Can you point to any existing real-world case where someone wrote this?

Neme12 replied:
> Without additional work, the blob will contain everything

That could be easily changed so that the blob would only contain the sliced array. I understand that this might not be super common, but my concern is that *since this is supported for the byte case*, my assumption would be that it would be supported for the newly supported types as well. It might be better to be consistent.

Neme12 replied:
Ah I see. Might be worth adding a comment saying that the case when `lengthForConstructor != elementCount` is only necessary for utf-8 literals, which is only needed for bytes.

AlekseyTs commented:
/azp run

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -443,21 +488,30 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
             if (elementCount == 0)
             {
-                if (inPlace)
+                // The span is empty.  If this is in-place initialization, call the default ctor.
```
>// The span is empty.  If this is in-place initialization, call the default ctor. [](http://example.com/codeflow?start=16&length=81)

Is this code path covered by tests for the new element types that we now can handle?

AlekseyTs commented:
Done with review pass (commit 3)

stephentoub commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -468,26 +522,77 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+                        // but tell the caller they can call this again with a null inPlaceTarget, at which point this
+                        // should be able to optimize the call.
+                        avoidInPlace = true;
+                        return false;
```
Yes, e.g. in https://github.com/dotnet/roslyn/pull/61414/files#diff-1f62aa94fc31edc1a6baa48cf8c18d07af7ae12ac0c04ed9edb2ff434fad03c2R853

AlekseyTs reviewed (dismissed):
LGTM (commit 4)

AlekseyTs commented:
@dotnet/roslyn-compiler For the second review.

jcouv commented on src/Compilers/Core/Portable/WellKnownMembers.cs:
```diff
@@ -992,6 +992,17 @@ static WellKnownMembers()
                 0,                                                                                                          // Arity
                     (byte)SignatureTypeCode.GenericTypeParameter, 0,                                                        // Field Signature
 
+                // System_Runtime_CompilerServices_RuntimeHelpers__CreateSpanRuntimeFieldHandle
```
The doc for new `CreateSpan` API should mention `uint` and probably also `bool` (although the compiler doesn't use it for that case)

Neme12 replied:
What about `nint` and `nuint`? Can this optimization not work for that as well?

stephentoub replied:
Those already aren't supported for arrays, e.g. RuntimeHelpers.InitializeArray.  It'd be hard to do that efficiently... the data is baked into the assembly, but their size can change based on that of the process into which the assembly is loaded.  So you'd have to pick one size and then at runtime adapt if it wasn't that.  We do the same for endianness, but big endian machines are rare (and aren't even supported with coreclr); 32-bit vs 64-bit is much more of a coin flip.

jcouv replied:
Chatted with @davidwrighton and there's a few things we need to follow up on regarding field RVA:

1. Is the compiler emitting properly aligned blobs?
David thinks the compiler always produces blobs aligned with 8 bytes (ie. a bit inefficient, but safe).
We'll need to double-check that, and fix if not the case.

2. System.Reflection.Metadata
This assembly is used by the compiler to emit PE. But it used to have a bug for scenarios with blobs aligned with 8 bytes.
This was fixed in January (PR https://github.com/dotnet/runtime/pull/63305), but we need to make sure that roslyn is using an up-to-date version of this binary, which is apparently a bit complicated. I will follow-up with Jared on Monday.

3. IL rewriters 
IL rewriters need to maintain alignment. We're ignoring CCI, so that leave Cecil scenarios.
Cecil was updated (PR https://github.com/mono/cecil/pull/33) to maintain alignment, but that design relies on the compiler tagging the field with a "pack" value (a number, such as 1, 2, 4, 8, ...., part of the field layout details).
We'll want the compiler to emit the "pack" value.
The trimmer uses Cecil and is part of the SDK, so users will automatically have this fix as part of their .NET 7 build.
But any customers that have custom trimmers based on Cecil would run into issues if their Cecil is not up-to-date. We could add some documentation in breaking changes [doc](https://github.com/dotnet/roslyn/blob/main/docs/compilers/CSharp/Compiler%20Breaking%20Changes%20-%20DotNet%207.md) as mitigation.
In terms of failure mode, the `CreateSpan` API currently detects that the blob is not properly aligned and will explicitly fail.

jaredpar replied:
>  we need to make sure that roslyn is using an up-to-date version of this binary, which is apparently a bit complicated. 

Yeah that will be a bit complicated. The fix went into `main` which means it will appear in the 7.0.0.0 version of SRM. That releases as a part of `net7.0` but we need it in the compiler, at RTM versions, before RTM date. Updating SRM or SCI is always "fun" as we have to change the VS versions and that is a bit of a hassle. 

Is there any chance of this being back ported to 6.0.0.0 servicing? That is the easiest route as it gives us the most wiggle room for getting everything inserted.

jcouv replied:
@jaredpar It looks like the SRM fix was backported to .NET 6 (PR https://github.com/dotnet/runtime/pull/64559) :-)

stephentoub replied:
> We'll want the compiler to emit the "pack" value.

I've pushed a commit that does so.  Let me know if it's what you had in mind.

stephentoub replied:
I added a src assert and test for (1) as well.

jcouv replied:
@stephentoub Quick update. From chat with Jared, we're going to wait for .NET 7 to RTM, then it should be relatively easy to insert RTM bits of SRM into VS.
I'm going to mark the PR as draft for now to make room in our review queue. Can you re-activate and ping after .NET 7 RTMs?

jcouv replied:
It looks like the work to insert newer SRM bits into VS is under way, although it might be two steps (6.0.1 first then 7). 6.0.1 would be sufficient for this PR. Here's the [VS PR to track](https://dev.azure.com/devdiv/DevDiv/_git/VS/pullrequest/434270).

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadOnlySpanConstructionTest.cs:
```diff
@@ -73,66 +74,70 @@ .maxstack  2
   IL_0013:  stloc.1
   IL_0014:  ldloca.s   V_0
-  IL_0016:  call       ""int System.ReadOnlySpan<byte>.Length.get""
+  IL_0016:  call       ""int System.ReadOnlySpan<{ type}>.Length.get""
```
>{ type} [](http://example.com/codeflow?start=49&length=7)

nit: formatting here and two lines below

stephentoub replied:
Grr, VS auto-formatting.  Thanks, will fix.

davidwrighton reviewed (approved):
I've reviewed the addition of use of Pack value, not anything else.

stephentoub commented on src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs:
```diff
@@ -69,7 +69,7 @@ internal sealed class PrivateImplementationDetails : DefaultTypeDef, Cci.INamesp
         // field types for different block sizes.
         private ImmutableArray<Cci.ITypeReference> _orderedProxyTypes;
-        private readonly ConcurrentDictionary<uint, Cci.ITypeReference> _proxyTypes = new ConcurrentDictionary<uint, Cci.ITypeReference>();
+        private readonly ConcurrentDictionary<(uint Size, ushort Alignment), Cci.ITypeReference> _proxyTypes = new ConcurrentDictionary<(uint Size, ushort Alignment), Cci.ITypeReference>();
```
@jcouv, note that an alternative to most of the changes in this commit would be always setting pack to 8 for anything that's not 1/2/4/8 bytes in length.  If you'd prefer that route, let me know.

stephentoub replied:
Also, it wasn't clear to me how the compiler is avoiding bugs when the hash codes happen to collide.  Is the assumption that it simply won't ever happen because it's using a cryptographic hash and the chances of collision are super small?  If there's other logic for handling such collisions, I can also change the naming logic to just piggyback on that instead.

AlekseyTs commented on src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs:
```diff
@@ -140,38 +140,28 @@ internal void Freeze()
 
-            return new ExplicitSizeStruct(size, this, _systemValueType);
+            return _mappedFields.GetOrAdd((data, alignment), key =>
+                new MappedField(GenerateDataFieldName(key.Data, key.Alignment), this, type, key.Data));
```
>new MappedField(GenerateDataFieldName(key.Data, key.Alignment), this, type, key.Data)) [](http://example.com/codeflow?start=16&length=86)

Does offset of this field in private implementation detail type make any difference?

AlekseyTs replied:
It looks like this field is static, so it is not a subject to layout within a type instance.

jcouv replied:
I'll let @davidwrighton confirm, but my understanding is that the offset of the field doesn't matter. What matters is the alignment and .pack flag on the blob that the field points to.

stephentoub replied:
> It looks like this field is static, so it is not a subject to layout within a type instance.

The .pack is for the type of the field (not the type containing the field), and according to the docs "The alignment of the type is the size of its largest element (1, 2, 4, 8, etc., bytes) or the specified packing size, whichever is smaller."

tannergooding commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -526,6 +528,8 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
                     return false;
                 }
 
+                Debug.Assert(specialElementType.SizeInBytes() is 1 or 2 or 4 or 8);
```
Are there no plans for `decimal`, which has a well-defined size and layout and which has constant support in other places?

Or is this possibly a "future consideration"?

jcouv commented on src/Compilers/Core/Portable/CodeGen/ILBuilderEmit.cs:
```diff
@@ -86,7 +86,7 @@ internal void EmitArrayBlockInitializer(ImmutableArray<byte> data, SyntaxNode sy
 
             // map a field to the block (that makes it addressable via a token)
-            var field = module.GetFieldForData(data, syntaxNode, diagnostics);
+            var field = module.GetFieldForData(data, alignment: 1, syntaxNode, diagnostics);
```
Just to confirm, the reason we're using alignment=1 here is because `System_Runtime_CompilerServices_RuntimeHelpers__InitializeArrayArrayRuntimeFieldHandle` doesn't care about alignment, correct?
Similarly, in next method, because of `ILOpCode.Cpblk`?

stephentoub replied:
Yes.  I can change it if you'd like, but this will likely cause many tests to need to be updated because doing so will change the generated field name.  If we want to go with https://github.com/dotnet/roslyn/pull/61414#discussion_r890169844 instead, it'd be easier to make this change for consistency even if it's not needed.

jcouv replied:
I'm fine to leave alignment=1 for scenarios that don't need special alignment. Maybe add a comment and add a test that captures that behavior explicitly.

jcouv commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -468,26 +524,79 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+
+                    // ldtoken <PrivateImplementationDetails>...
+                    // call ReadOnlySpan<elementType> RuntimeHelpers::CreateSpan<elementType(fldHandle)
+                    var field = _builder.module.GetFieldForData(data, (ushort)specialElementType.SizeInBytes(), wrappedExpression.Syntax, _diagnostics);
```
>(ushort)specialElementType.SizeInBytes(), [](http://example.com/codeflow?start=70&length=41)

nit: Consider naming this argument (`alignment: `)

jcouv commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -468,26 +524,79 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+                    }
+
+                    // ldtoken <PrivateImplementationDetails>...
+                    // call ReadOnlySpan<elementType> RuntimeHelpers::CreateSpan<elementType(fldHandle)
```
typo: `CreateSpan<elementType>` (missing `>`)

jcouv commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -468,26 +524,79 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+                        EmitAddress(inPlaceTarget, Binder.AddressKind.Writeable);
+                    }
+
+                    _builder.EmitArrayBlockFieldRef(data, (ushort)specialElementType.SizeInBytes(), wrappedExpression.Syntax, _diagnostics);
```
I didn't understand why we're passing an alignment value here. This branch doesn't use `CreateSpan` API, and it seems like you opted to not align for scenarios that don't technically need it.

stephentoub replied:
specialElementType here is always `byte`/`sbyte`/`bool`, and thus is always 1.  I can just change it to hardcode it.

jcouv replied:
I think we can hardcode it (ie. remove parameter from `EmitArrayBlockFieldRef`).

> specialElementType here is always byte/sbyte/bool, and thus is always 1. 
Oh, I see that now. Thanks
Consider adding an assertion in this branch.

stephentoub commented on src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs:
```diff
@@ -140,38 +140,28 @@ internal void Freeze()
+                    1 when _systemInt8Type is not null => _systemInt8Type,
+                    2 when _systemInt16Type is not null => _systemInt16Type,
+                    4 when _systemInt32Type is not null => _systemInt32Type,
+                    8 when _systemInt64Type is not null => _systemInt64Type,
```
> those special types don't have an Alignment value

Those special types are Byte, Short, Int32, and Int64, which by definition need to be aligned according to the runtime's requirements.  Am I misunderstanding the question or what these "special types" get resolved to?

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadOnlySpanConstructionTest.cs:
```diff
@@ -719,5 +724,253 @@ .maxstack  2
+                yield return new object[] { withCtor, "MyColor_Int32", "MyColor_Int32.Red, MyColor_Int32.Blue", "ldind.i4", "long <PrivateImplementationDetails>.1B03AB083D0FB41E44D480F48D5BBA181C623C0594BDA1AA8EA71A3B67DBF3B14" };
+                yield return new object[] { withCtor, "MyColor_UInt32", "MyColor_UInt32.Red, MyColor_UInt32.Blue", "ldind.u4", "long <PrivateImplementationDetails>.1B03AB083D0FB41E44D480F48D5BBA181C623C0594BDA1AA8EA71A3B67DBF3B14" };
+                yield return new object[] { withCtor, "MyColor_Int64", "MyColor_Int64.Red, MyColor_Int64.Blue", "ldind.i8", "<PrivateImplementationDetails>.__StaticArrayInitTypeSize=16 <PrivateImplementationDetails>.F7548C023E431138B11357593F5CCEB9DD35EB0B0A2041F0B1560212EEB6F13E8" };
+                yield return new object[] { withCtor, "MyColor_UInt64", "MyColor_UInt64.Red, MyColor_UInt64.Blue", "ldind.i8", "<PrivateImplementationDetails>.__StaticArrayInitTypeSize=16 <PrivateImplementationDetails>.F7548C023E431138B11357593F5CCEB9DD35EB0B0A2041F0B1560212EEB6F13E8" };
```
What we care to verify on the field RVA question is that blobs for private implementation detail fields are emitted (1) with proper alignment and (2) with proper .pack flag.
The current tests don't really show that. Let's try to get that coverage. 
I'm not exactly sure how to test that, I'll dig around to see how to demonstrate we're doing the correct thing.

stephentoub replied:
> The current tests don't really show that.

Yes, that wasn't the intent of these.  I can add additional tests covering .pack.

AlekseyTs commented:
Do we still have to meet "The Rva of the field must be aligned on a natural boundary of type T" requirement?

jcouv commented on src/Compilers/Core/Portable/CodeGen/ILBuilderEmit.cs:
```diff
@@ -119,10 +120,10 @@ internal void EmitStackAllocBlockInitializer(ImmutableArray<byte> data, SyntaxNo
         }
 
-        internal void EmitArrayBlockFieldRef(ImmutableArray<byte> data, SyntaxNode syntaxNode, DiagnosticBag diagnostics)
+        internal void EmitArrayBlockFieldRef(ImmutableArray<byte> data, ushort alignment, SyntaxNode syntaxNode, DiagnosticBag diagnostics)
```
The alignment parameter can be removed (only caller hardcodes 1)

jcouv commented on src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs:
```diff
@@ -2321,6 +2321,8 @@ private void PopulateFieldRvaTableRows(BlobBuilder mappedFieldDataWriter)
                 }
 
                 int offset = mappedFieldDataWriter.Count;
+                Debug.Assert(offset % ManagedPEBuilder.MappedFieldDataAlignment == 0, "Expected last write to end at alignment boundary");
```
nit: consider asserting with specific value (8)

stephentoub replied:
Using the named const seems better to me.  But since you asked, I added an assert that the named const is 8 as well.

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadOnlySpanConstructionTest.cs:
```diff
@@ -719,5 +724,336 @@ .maxstack  2
+}";
+
+            // All blob RVAs are 8-byte aligned, regardless of the data type.
+            string byteBlobRva = ExecutionConditionUtil.IsWindows ? "I_00002AE0" : "I_00002AE0"; // TODO: fix Unix values
```
We can't add new TODO comments in the codebase. For something that needs to be addressed later, we file new issues and put link in code. But this looks like something you intend to fix within this PR.

stephentoub replied:
> But this looks like something you intend to fix within this PR.

Yes, I'm using CI to get all the values.

jcouv reviewed (commented):
Done with review pass (iteration 7). Only minor comments left

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadOnlySpanConstructionTest.cs:
```diff
@@ -719,5 +726,288 @@ .maxstack  2
+                Assert.True(long.TryParse(m.Groups[1].Value, NumberStyles.HexNumber, null, out long rva), $"Expected {m.Value} to parse as hex long.");
+                Assert.True(rva % expectedAlignment == 0, $"Expected RVA {rva:X8} to be {expectedAlignment}-byte aligned.");
+            });
+        }
```
Code and tests look good to me, but we should still chase that SRM version question. Let's sync with Jared on this tomorrow.

pentp commented:
> > > Would be nice if ReadOnlySpan span = "text"; would also be optimized, since it's a much more convenient syntax for strings and would be analogous to the utf8 string feature.
> > 
> > 
> > What is it you'd hope it would do it's not already doing? [SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHunGAlGNgAmAeQB2AGwCeAZQAO2MQB4wAC2xQAfAwCy5ABQBKBgF5tAIgwwEGcwG4aAXyA==)
> 
> I thought the runtime materialized the string and created a span over it then, but it seems I was wrong despite the IL using `ldstr`.

Unfortunately that is what actually happens: [SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACY8pJ0hgYQYG8aH+mjZqwBKMbABMA8gDsANgE8AygAdsMgDxgAFtigA+BgFlyACgCUDALyGARBhgIMtgNx8B9Jiy8AGY6QtrOwcnVxoAXyA===) (same indirect pointer is used for both methods)

Ideally the ROS would point directly to the string data from the binary, without materializing the string at all. To avoid data duplication between identical strings and ROS, it would need to point to the UserString heap, but that is probably not possible directly using Roslyn and should actually be handled by the runtime/JIT?

jaredpar commented:
VS has now updated SRM to a version that has all the bug fixes necessary for this change 

https://dev.azure.com/devdiv/DevDiv/_git/VS/pullrequest/434270

stephentoub commented:
Thanks, @jaredpar and @jcouv.

I've updated the PR and marked it as ready for review.  This includes the change we talked about, where we use CreateSpan if it's available and otherwise use a cached array.

stephentoub commented:
Test failures are https://github.com/dotnet/roslyn/issues/65566.

jcouv commented on src/Compilers/CSharp/Portable/CodeGen/EmitConversion.cs:
```diff
@@ -74,7 +74,7 @@ private void EmitReadOnlySpanFromArrayExpression(BoundReadOnlySpanFromArray expr
                          "only special kinds of conversions involving ReadOnlySpan may be handled in emit");
 
-            if (!TryEmitReadonlySpanAsBlobWrapper(typeTo, operand, used, inPlace: false))
+            if (!TryEmitReadonlySpanAsBlobWrapper(typeTo, operand, used, inPlaceTarget: null, out _))
```
>out _ [](http://example.com/codeflow?start=94&length=5)

nit: consider naming discarded `out` arguments to `TryEmitReadonlySpanAsBlobWrapper` (here and elsewhere)

jcouv commented on src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs:
```diff
@@ -140,38 +140,46 @@ internal void Freeze()
+            // Call sites will lazily instantiate the array to cache in this field, rather than forcibly instantiating
+            // all of them when the private implementation details class is first used.
+            return _mappedFields.GetOrAdd((data, alignment, IsArray: true), key =>
+                new MappedField(GenerateDataFieldName(key.Data, key.Alignment, isArray: true), this, arrayType, default));
```
>default [](http://example.com/codeflow?start=112&length=7)

nit: let's name the `default` argument (`block:`)

jcouv commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -371,66 +371,113 @@ private static bool IsMultidimensionalInitializer(ImmutableArray<BoundExpression
+        /// <param name="wrappedExpression">The expression being wrapped in a span.</param>
+        /// <param name="used">true if the result of the expression is used; false if it's required only for its side effects.</param>
+        /// <param name="inPlaceTarget">A non-null expression if the construction is initializing an existing local in-place; otherwise, null.</param>
+        /// <param name="avoidInPlace">An output Boolean indicating whether a caller trying to perform in-place initialization should instead prefer to assign the local to a new value.</param>
```
nit: consider mentioning an example ("(for instance when we can use `CreateSpan` optimization, then the caller should call again with `null` `inPlaceTarget`)").

jcouv commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -371,66 +371,113 @@ private static bool IsMultidimensionalInitializer(ImmutableArray<BoundExpression
-                //     the issue is that metadata stores blobs in little-endian format
-                //     so anything that is larger than one byte will be incorrect on a big-endian machine
-                //     With additional runtime support it might be possible, but not yet.
-                //     See: https://github.com/dotnet/corefx/issues/26948 for more details
```
Looking at this removed comment made me think that the endian-ness behavior should be mentioned in the `CreateSpan` doc. I assume that on certain platforms it will copy and rectify endian-ness, while on the more common platform no copy/rectification is needed.

jcouv commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -371,66 +371,113 @@ private static bool IsMultidimensionalInitializer(ImmutableArray<BoundExpression
+                // Get the array element type.  This optimization is only supported for core primitive types.
+                arrayType = (ArrayTypeSymbol)ac.Type;
+                specialElementType = arrayType.ElementType.EnumUnderlyingTypeOrSelf().SpecialType;
+                switch (specialElementType)
```
nit: this method is getting rather long. Consider extracting a local function and doing `if (!isTypeAllowedInBlobWrapper(specialElementType) return false;` here.

jcouv commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -371,66 +371,113 @@ private static bool IsMultidimensionalInitializer(ImmutableArray<BoundExpression
+
             if (start is null != length is null)
             {
+                // We can only handle the case where there's a start expression iff there's a length expression.
```
Based on the assertion at the top of the method, we cannot enter this condition. Instead of assertion plus this `if` statement, let's just do one `if` statement with `throw new ExceptionUtilities.Unreachable();` at the top of the method.

jcouv commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -445,51 +492,154 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+
+            if (inPlaceTarget is not null)
+            {
+                // We can use RuntimeHelpers.CreateSpan / cached array but not for in-place initialization. Fail to optimize,
```
>/ [](http://example.com/codeflow?start=56&length=1)

Let's use a word instead of a symbol for clarity (also applies above)

jcouv commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -445,51 +492,154 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+            }
+            Debug.Assert(!rosArrayCtor.HasUnsupportedMetadata);
+
+            ushort alignment = (ushort)specialElementType.SizeInBytes();
```
nit: The below emit sequence could use some comments (some pseudo-code)

jcouv commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -445,51 +492,154 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+                // CreateSpan was available. Use it.
+                // ldtoken <PrivateImplementationDetails>...
+                // call ReadOnlySpan<elementType> RuntimeHelpers::CreateSpan<elementType>(fldHandle)
+                var field = _builder.module.GetFieldForData(data, alignment: (ushort)specialElementType.SizeInBytes(), wrappedExpression.Syntax, _diagnostics.DiagnosticBag!);
```
>_diagnostics.DiagnosticBag! [](http://example.com/codeflow?start=145&length=27)

The guideline on nullability suppressions is to only use `!` if the method already checks, but the compiler wasn't able to analyze/track the information. Otherwise, an assertion is preferred.
Comment applies to multiple locations in this method.

jcouv commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -371,66 +371,113 @@ private static bool IsMultidimensionalInitializer(ImmutableArray<BoundExpression
+            // and for platforms that lack CreateSpan, as a span that wraps a lazily-initialized array:
+            //     new ReadOnlySpan<T>(PrivateImplementationDetails.ArrayField ??= new T[] { ... })
+            // For non-constant data, unsupported primitive types, and other variations, this optimization will fail
+            // and the method will return false indicating that no code was emitted.
```
Thanks a lot for the explanations (here and below in this method). It really helps to follow this method :-)

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadOnlySpanConstructionTest.cs:
```diff
@@ -908,5 +915,296 @@ .maxstack  4
+            });
+        }
+
+        private const string RuntimeHelpersCreateSpan = @"
```
New tests use this `RuntimeHelpers` API, but old tests (currently using older TFMs) don't have this API. At some point, the old tests that aren't locked to a specific TFM will be updated to new TFMs that includes this new API. 
It would be good if we can make it easier for the person updating the TFMs, so they don't have to figure out how to update or fix many tests.

I'm thinking we should update existing tests to make the API missing explicitly, with `compilation.MakeMemberMissing(WellKnownMember.System_Runtime_CompilerServices_RuntimeHelpers__CreateSpanRuntimeFieldHandle);`)

jcouv replied:
Gentle ping

stephentoub replied:
I added it into the tests I added that expect CreateSpan to not exist. Which other tests are you referring to?

jcouv commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -371,66 +371,113 @@ private static bool IsMultidimensionalInitializer(ImmutableArray<BoundExpression
-
+            Debug.Assert(!rosPointerCtor.HasUnsupportedMetadata);
+
+            // The purpose of this optimization is to replace a BoundArrayCreation with better code generation.
```
It would be good to mention the UTF8 scenario which you mentioned offline. That motivates why this method should handle start/length scenarios, albeit only when element size is 1.

jcouv commented on src/Compilers/Core/Portable/CodeGen/ITokenDeferral.cs:
```diff
@@ -13,7 +13,12 @@ internal interface ITokenDeferral
+
+        /// <summary>Gets a field that may be used to lazily cache an array created to store the specified data.</summary>
+        /// <remarks>This is used to cache an array created with the exact same arguments passed to <see cref="GetFieldForData"/>.</remarks>
+        Cci.IFieldReference GetArrayCachingFieldForData(ImmutableArray<byte> data, ushort alignment, Cci.IArrayTypeReference arrayType, SyntaxNode syntaxNode, DiagnosticBag diagnostics);
```
>ushort alignment [](http://example.com/codeflow?start=83&length=16)

This `alignment` parameter is confusing. Do we really need it? 
The only call-site currently passes the element size, which doesn't seem relevant when creating a field of array type. 
Consider removing it and hard-coding an `alignment: 1` where this value is eventually used (call to `GenerateDataFieldName` in `CreateArrayCachingField`).

jcouv commented on src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs:
```diff
@@ -2317,6 +2317,9 @@ private void PopulateFieldRvaTableRows(BlobBuilder mappedFieldDataWriter)
 
                 int offset = mappedFieldDataWriter.Count;
+                Debug.Assert(offset % ManagedPEBuilder.MappedFieldDataAlignment == 0, "Expected last write to end at alignment boundary");
+                Debug.Assert(ManagedPEBuilder.MappedFieldDataAlignment == 8, "Expected alignment to be 8");
```
nit: this may be a good location to document the fact that we emit RVAs as multiples of 8, but packingSize/.pack/Alignment emitted by `PopulateClassLayoutTableRows` is finer-grained. Alternatively, in `PopulateClassLayoutTableRows` could be good too.

jcouv commented on src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs:
```diff
@@ -140,38 +140,46 @@ internal void Freeze()
 
-        internal Cci.IFieldReference CreateDataField(ImmutableArray<byte> data)
+        /// <summary>
+        /// Gets a field that can be used to cache an array allocated to store data from a corresponding <see cref="CreateDataField"/> call.
```
nit: I thought this comment was a bit confusing. `CreateArrayCachingField` is used when we don't end up using a blob stored via `CreateDataField`. So I don't think `CreateDataField` should be mentioned here.

jcouv commented on src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs:
```diff
@@ -287,7 +295,24 @@ internal static string GenerateDataFieldName(ImmutableArray<byte> data)
+            // just use the hex value of the data hash.  For other alignments, tack on a '2', '4', or '8'
+            // accordingly.  As every byte will yield two chars, the odd number of chars used for 2/4/8
+            // alignments will never produce a name that conflicts with names for an alignment of 1.
+            Debug.Assert(alignment is 1 or 2 or 4 or 8);
```
Does this assertion hold? Can't `alignment` be `3` here?
If the assertion holds, then it could be tightened by handling the `1` case explicitly and throwing unreachable exception for the switch's default case below.

jcouv replied:
nit: Consider handling the `1` case explicitly in `switch` below and throwing unreachable exception for the switch's default case. Then we don't need this assertion.

jcouv commented on src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs:
```diff
@@ -318,20 +358,22 @@ public int Compare(SynthesizedStaticField? x, SynthesizedStaticField? y)
+        internal ExplicitSizeStruct(uint size, ushort alignment, PrivateImplementationDetails containingType, Cci.ITypeReference sysValueType)
         {
             _size = size;
+            _alignment = alignment;
```
This also may be a useful place to assert on alignment allowed values.

jcouv commented on src/Compilers/Core/Portable/CodeGen/ILBuilderEmit.cs:
```diff
@@ -109,7 +109,8 @@ internal void EmitStackAllocBlockInitializer(ImmutableArray<byte> data, SyntaxNo
             {
-                var field = module.GetFieldForData(data, syntaxNode, diagnostics);
+                // alignment==1 as there's no special need for alignment (.pack) here.
+                var field = module.GetFieldForData(data, alignment: 1, syntaxNode, diagnostics);
```
:memo: We have a single caller that comes into this branch (`emitInitBlock: false`) and it only calls this method when `elementType.SpecialType.SizeInBytes() == 1`. So `alignment: 1` is correct.
Similarly with `EmitArrayBlockFieldRef` below, we're fine with `alignment: 1` because it's only ever called when `specialElementType.SizeInBytes() == 1`.

jcouv commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -445,51 +492,154 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+                if (createSpan.GetUseSiteInfo().DiagnosticInfo?.Severity == DiagnosticSeverity.Error)
+                {
+                    // Something went wrong.
+                    return false;
```
Were you able to hit this case? (an Obsolete attribute will probably do it)

stephentoub replied:
I wasn't able to trigger this.  I tried putting an ObsoleteAttribute on CreateSpan, even with error: true, and it didn't hit this path.

AlekseyTs replied:
> I wasn't able to trigger this.  I tried putting an ObsoleteAttribute on CreateSpan, even with error: true, and it didn't hit this path.

See unit test ```StaticFieldIsUsedForSpanCreatedFromArrayWithInitializer_02``` that covers scenario for the ```rosPointerCtor```

stephentoub replied:
Doesn't that test use MakeMemberMissing to make it so that GetWellKnownTypeMember returns null rather than there being an error at the use site?  Or have I misunderstood?

AlekseyTs replied:
It also covers scenario when method is there and would be found by compilation (as you do that right now), but is bad and should not be used, Read the comment at the beginning of the unit-test.

AlekseyTs replied:
It also covers scenario when method is there and would be found by compilation (as you do that right now), but is bad and should not be used, Read the comment at the beginning of the unit-test.

jcouv commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -445,51 +492,154 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+            //     new ReadOnlySpan<T>(PrivateImplementationDetails.ArrayField ??= RuntimeHelpers.InitializeArray(new int[Length], PrivateImplementationDetails.DataField));
+
+            var rosArrayCtor = (MethodSymbol?)Binder.GetWellKnownTypeMember(_module.Compilation, WellKnownMember.System_ReadOnlySpan_T__ctor_Array, _diagnostics, syntax: wrappedExpression.Syntax, isOptional: true);
+            if (rosArrayCtor is null)
```
Should we do a similar `GetUseSiteInfo()` check for `WellKnownMember.System_ReadOnlySpan_T__ctor_Array` as above for `WellKnownMember.System_ReadOnlySpan_T__ctor_Array`?

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadOnlySpanConstructionTest.cs:
```diff
@@ -908,5 +915,296 @@ .maxstack  4
+  IL_0014:  stelem.i2
+  IL_0015:  ldc.i4.{start}
+  IL_0016:  ldc.i4.{length}
+  IL_0017:  newobj     ""System.ReadOnlySpan<char>..ctor(char[], int, int)""
```
Let's also include a negative case, where the element size is not 1 and start/length arguments are provided (then the optimization doesn't kick in).

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadOnlySpanConstructionTest.cs:
```diff
@@ -908,5 +915,296 @@ .maxstack  4
+        [InlineData(1, 2)]
+        [InlineData(1, 3)]
+        [InlineData(0, 4)]
+        public void CreateSpanIsNotUsedForSpanOfNonSize1Types_NonFullLength(int start, int length)
```
Is the test name correct? This test uses a size-1 type.

stephentoub replied:
It uses char, which is 2 bytes.

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadOnlySpanConstructionTest.cs:
```diff
@@ -908,5 +915,296 @@ .maxstack  4
+  // Code size       11 (0xb)
+  .maxstack  1
+  IL_0000:  ldtoken    ""{ldtokenArg}""
+  IL_0005:  call       ""System.ReadOnlySpan<{type}> System.Runtime.CompilerServices.RuntimeHelpers.CreateSpan<{type}>(System.RuntimeFieldHandle)""
```
This test covers all the positive types intended for the optimization. Consider adding a similar test with a type that can be used for constants but would not trigger the CreateSpan optimization (such as `decimal`).

stephentoub replied:
`decimal` is covered by the CreateSpanIsNotUsedForNonSupportedTypes later on.

jcouv commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -445,51 +492,154 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+                // We can use RuntimeHelpers.CreateSpan / cached array but not for in-place initialization. Fail to optimize,
+                // but tell the caller they can call this again with a null inPlaceTarget, at which point this
+                // should be able to optimize the call.
+                avoidInPlace = true;
```
Do we have a test that hits this branch and illustrates this behavior? (Optimization will kick in but the in-place attempt will be dropped in favor of an assignment attempt with the new optimization)

jcouv replied:
Gentle ping on question

stephentoub replied:
It's hit in one of the new tests I added.

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadOnlySpanConstructionTest.cs:
```diff
@@ -908,5 +915,296 @@ .maxstack  4
+                Assert.True(long.TryParse(m.Groups[1].Value, NumberStyles.HexNumber, null, out long rva), $"Expected {m.Value} to parse as hex long.");
+                Assert.True(rva % expectedAlignment == 0, $"Expected RVA {rva:X8} to be {expectedAlignment}-byte aligned.");
+            });
+        }
```
A couple of additional test ideas:
1. a compilation with two blobs with same element type, either with same data or different data (to show the sharing or non-sharing)
2. a compilation with two blobs with same hashcodes but different element types (if `CryptographicHashProvider.ComputeSourceHash` allows such an example to be constructed easily)

jcouv reviewed (commented):
Done with review pass (iteration 5)

MichalPetryka commented:
Is there something planned for specifying custom alignment for RVA static tables for use in SIMD code? In such cases Vector width alignment is beneficial for performance.

stephentoub commented:
> Is there something planned for specifying custom alignment for RVA static tables for use in SIMD code? In such cases Vector width alignment is beneficial for performance.

Not to my knowledge, and certainly not as part of this work.

stephentoub commented:
Thanks for the review, @jcouv.  I've addressed the feedback.

jcouv reviewed (approved):
LGTM Thanks (iteration 8)

jcouv commented:
@AlekseyTs @dotnet/roslyn-compiler for second review. Thanks

AlekseyTs commented on src/Compilers/Core/Portable/CodeGen/ILBuilderEmit.cs:
```diff
@@ -109,7 +109,8 @@ internal void EmitStackAllocBlockInitializer(ImmutableArray<byte> data, SyntaxNo
             else
             {
-                var field = module.GetFieldForData(data, syntaxNode, diagnostics);
+                // alignment==1 as emitInitBlock is false only if the element type for the stackalloc has SizeInBytes == 1.
```
>// alignment==1 as emitInitBlock is false only if the element type for the stackalloc has SizeInBytes == 1. [](http://example.com/codeflow?start=16&length=107)

Rather than making this claim in a comment, consider inlining the code at the call site. It looks like there is only one call site per distinct ```emitInitBlock``` value. Then code will be explicitly guarded by the size check and we wouldn't need to worry about violating this assumption in the future,

AlekseyTs commented on src/Compilers/Core/Portable/CodeGen/ILBuilderEmit.cs:
```diff
@@ -121,8 +122,9 @@ internal void EmitStackAllocBlockInitializer(ImmutableArray<byte> data, SyntaxNo
-            // map a field to the block (that makes it addressable)
-            var field = module.GetFieldForData(data, syntaxNode, diagnostics);
+            // Map a field to the block (that makes it addressable).
+            // alignment==1 as this is only ever used when the element type has a SizeInBytes == 1.
```
>// alignment==1 as this is only ever used when the element type has a SizeInBytes == 1. [](http://example.com/codeflow?start=12&length=87)

Similar suggestion to inline the code at the only call site

AlekseyTs commented on src/Compilers/Core/Portable/CodeGen/ITokenDeferral.cs:
```diff
@@ -13,7 +13,12 @@ internal interface ITokenDeferral
+
+        /// <summary>Gets a field that may be used to lazily cache an array created to store the specified data.</summary>
+        /// <remarks>This is used to cache an array created with the data passed to <see cref="GetFieldForData"/>.</remarks>
+        Cci.IFieldReference GetArrayCachingFieldForData(ImmutableArray<byte> data, Cci.IArrayTypeReference arrayType, SpecialType elementType, SyntaxNode syntaxNode, DiagnosticBag diagnostics);
```
>SpecialType elementType [](http://example.com/codeflow?start=118&length=23)

Can't we get this from ```arrayType```?

stephentoub replied:
I wanted to do that but couldn't figure out how.  There's an IArrayTypeReference.GetElementType method but it requires an EmitContext instance to be passed in, and I wasn't sure where to get that from.  If you know, I'd be happy to change it.

AlekseyTs replied:
>I wanted to do that but couldn't figure out how.  There's an IArrayTypeReference.GetElementType method but it requires an EmitContext instance to be passed in, and I wasn't sure where to get that from.  If you know, I'd be happy to change it.

```CommonPEModuleBuilder``` creates one on the fly in several places.

stephentoub replied:
The two such occurrences I saw do:
```C#
new EmitContext(this, syntaxNode, diagnostics, metadataOnly: false, includePrivateMembers: true)
```
If it's valid here to similarly assume metadataOnly=false and includePrivateMembers=true (I don't know what the implications of that are), then I should be able to take this approach if it's considered better than just passing in the SpecialType we already have.

AlekseyTs replied:
I think ```includePrivateMembers=false``` is meaningful on;y when ```metadataOnly=true```., And since we are emitting IL (we wouldn't need to call the helpers otherwise, I think), ```metadataOnly=false``` feels appropriate

AlekseyTs replied:
> if it's considered better than just passing in the SpecialType we already have.

We should avoid passing redundant data around. Also, we should avoid creating a situation where one can pass inconsistent input by mistake. I.e. pass ```SpecialType``` that doesn't match array's element type, for example.

AlekseyTs commented on src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs:
```diff
@@ -140,38 +140,64 @@ internal void Freeze()
+            // Call sites will lazily instantiate the array to cache in this field, rather than forcibly instantiating
+            // all of them when the private implementation details class is first used.
+            return _mappedFields.GetOrAdd((data, 1, elementType), key =>
+                new MappedField(GenerateDataFieldName(key.Data, key.Alignment, key.ElementType), this, arrayType, block: default));
```
>MappedField [](http://example.com/codeflow?start=20&length=11)

It is quite possible that I am missing something, but it feels inappropriate to to use this type for a field that is not associated with a blob.

stephentoub replied:
I can change the comment on the type (e.g. "Definition of a simple field mapped to a metadata block or that stores an array used to cache such a metadata block") or change the name of the type.  But it seemed like unnecessary complication to duplicate the type and the collection and and their handling; this is just expanding the use of something that was originally there for a single purpose and now that purpose is expanded.  If you really believe it'd be better to duplicate it all, though, it's your code base, I of course can.

AlekseyTs replied:
I don't think we should be reusing that type. Both from the clarity and instance size perspective.

stephentoub replied:
I don't agree, but it's your code base, so I won't argue.

AlekseyTs commented on src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs:
```diff
@@ -140,38 +140,64 @@ internal void Freeze()
+            // Create a dedicated mapped field for the array type, separate from the data that'll be stored into that array.
+            // Call sites will lazily instantiate the array to cache in this field, rather than forcibly instantiating
+            // all of them when the private implementation details class is first used.
+            return _mappedFields.GetOrAdd((data, 1, elementType), key =>
```
>1 [](http://example.com/codeflow?start=49&length=1)

I am not comfortable with usage of dummy value just so we could share a dictionary.

AlekseyTs commented on src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs:
```diff
@@ -140,38 +140,64 @@ internal void Freeze()
+                    _ => new ExplicitSizeStruct(key.Size, key.Alignment, this, _systemValueType)
+                });
+
+            return _mappedFields.GetOrAdd((data, alignment, SpecialType.None), key =>
```
>SpecialType.None [](http://example.com/codeflow?start=60&length=16)

I am not comfortable with usage of dummy values just so we could share a dictionary.

AlekseyTs commented on src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs:
```diff
@@ -276,18 +302,48 @@ public Cci.IUnitReference GetUnit(EmitContext context)
+        /// <param name="alignment">The alignment of the data.</param>
+        /// <param name="elementType">NotPrimitive if this field will store the data directly; otherwise, the element primitive type if this is to be used for the array with that type.</param>
+        /// <returns></returns>
+        internal static string GenerateDataFieldName(ImmutableArray<byte> data, ushort alignment = 1, SpecialType elementType = SpecialType.None)
```
>GenerateDataFieldName [](http://example.com/codeflow?start=31&length=21)

We should split this function, share the common part in a reusable helper, and otherwise have logic specific to particular field kind.

AlekseyTs replied:
>ushort alignment = 1, SpecialType elementType = SpecialType.None [](http://example.com/codeflow?start=80&length=64)

Having default values here is error prone, Consumer must provide explicit values at the call site.

AlekseyTs commented on src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs:
```diff
@@ -140,38 +140,64 @@ internal void Freeze()
+        /// <param name="data">The data that will be used to initialize the field.</param>
+        /// <param name="arrayType">The type of the field, e.g. int[].</param>
+        /// <returns>The field to use to cache an array for this data and alignment.</returns>
+        internal Cci.IFieldReference CreateArrayCachingField(ImmutableArray<byte> data, Cci.IArrayTypeReference arrayType, SpecialType elementType)
```
>CreateArrayCachingField [](http://example.com/codeflow?start=37&length=23)

Is my understanding correct that we don't want to share this field across blobs with same content and alignment, but with different element types?

stephentoub replied:
This caches a T[], so it should be unique to the T and data contained.  Technically in the future we could likely relax that requirement slightly if it proved fruitful, and use the same array when the T was compatible, e.g. share an array for both uint and int with the same data, but that complication doesn't currently seem warranted.

AlekseyTs commented on src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs:
```diff
@@ -140,38 +140,64 @@ internal void Freeze()
+        /// <param name="data">The data that will be used to initialize the field.</param>
+        /// <param name="arrayType">The type of the field, e.g. int[].</param>
+        /// <returns>The field to use to cache an array for this data and alignment.</returns>
+        internal Cci.IFieldReference CreateArrayCachingField(ImmutableArray<byte> data, Cci.IArrayTypeReference arrayType, SpecialType elementType)
```
>elementType [](http://example.com/codeflow?start=135&length=11)

I assume we do not expect to get ```SpecialType.None``` here. Let's assert this fact.

AlekseyTs commented on src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs:
```diff
@@ -310,6 +366,21 @@ public int Compare(SynthesizedStaticField? x, SynthesizedStaticField? y)
+
+        private sealed class MappedFieldsEqualityComparer : EqualityComparer<(ImmutableArray<byte> Data, ushort Alignment, SpecialType ElementType)>
+        {
+            public static readonly MappedFieldsEqualityComparer Instance = new();
```
>new() [](http://example.com/codeflow?start=75&length=5)

Please avoid adding new instances of target typed `new`

AlekseyTs commented on src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs:
```diff
@@ -55,8 +56,7 @@ internal sealed class PrivateImplementationDetails : DefaultTypeDef, Cci.INamesp
         private ImmutableArray<SynthesizedStaticField> _orderedSynthesizedFields;
-        private readonly ConcurrentDictionary<ImmutableArray<byte>, MappedField> _mappedFields =
-            new ConcurrentDictionary<ImmutableArray<byte>, MappedField>(ByteSequenceComparer.Instance);
+        private readonly ConcurrentDictionary<(ImmutableArray<byte> Data, ushort Alignment, SpecialType ElementType), MappedField> _mappedFields = new(MappedFieldsEqualityComparer.Instance);
```
>new [](http://example.com/codeflow?start=147&length=3)

Please avoid adding new instances of target typed `new`

AlekseyTs commented on src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs:
```diff
@@ -490,8 +565,6 @@ internal sealed class MappedField : SynthesizedStaticField
         internal MappedField(string name, Cci.INamedTypeDefinition containingType, Cci.ITypeReference type, ImmutableArray<byte> block)
             : base(name, containingType, type)
         {
-            Debug.Assert(!block.IsDefault);
```
>Debug.Assert(!block.IsDefault); [](http://example.com/codeflow?start=12&length=31)

It feels like this assert should be restored

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -52,7 +53,7 @@ private void EmitArrayInitializers(ArrayTypeSymbol arrayType, BoundArrayInitiali
             {
                 ImmutableArray<byte> data = this.GetRawData(initExprs);
-                _builder.EmitArrayBlockInitializer(data, inits.Syntax, _diagnostics.DiagnosticBag);
+                _builder.EmitArrayBlockInitializer(data, alignment: 1, inits.Syntax, _diagnostics.DiagnosticBag); // alignment == 1 as there's no special need for alignment (.pack) here.
```
>alignment: 1 [](http://example.com/codeflow?start=57&length=12)

Could this default potentially prevent sharing of the blob for a scenario when we could use the "right" specific alignment?

stephentoub replied:
This could lead to not sharing a blob across an array and a span created with the same constant data, e.g.
```C#
long[] array = new long[] { 1, 2, 3, 4, 5, 6, 7, 8 };
ReadOnlySpan<long> span = new long[] { 1, 2, 3, 4, 5, 6, 7, 8 };
```
will end up with two distinct blobs. The flip side, however, is that changing this alignment for arrays could lead to a rewriter (e.g. illink) that respects the .pack field regressing size a bit because it'll start enforcing that larger alignment.  I didn't want to mess with that right now, as I don't think we know yet whether that's a good tradeoff.

AlekseyTs replied:
Consider capturing this information in a comment.

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -371,66 +372,121 @@ private static bool IsMultidimensionalInitializer(ImmutableArray<BoundExpression
-
-            if (ctor == null)
+            var rosPointerCtor = (MethodSymbol?)Binder.GetWellKnownTypeMember(_module.Compilation, WellKnownMember.System_ReadOnlySpan_T__ctor_Pointer, _diagnostics, syntax: wrappedExpression.Syntax, isOptional: true);
+            if (rosPointerCtor is null)
```
>rosPointerCtor [](http://example.com/codeflow?start=16&length=14)

Is this helper always needed? Can its absence prevent optimization that is still possible to do?

stephentoub replied:
We could fall back to using the array caching scheme if the `ReadOnlySpan<T>(T[])` constructor exists.  But I don't think it's worth complicating the compiler to try to optimize the case of one of the constructors existing but not the other.  Both constructors were both added to `ReadOnlySpan<T>` when it was initially released, in both .NET Core and in the System.Memory nuget package available for use downlevel.  We'd be optimizing for a case that's not meaningful and that would basically never be hit outside of the compiler's tests.

stephentoub commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -445,54 +501,192 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
 
+        /// <summary>Gets whether the element type of an array is appropriate for storing in a blob.</summary>
+        private static bool IsTypeAllowedInBlobWrapper(SpecialType type) =>
+            type switch
```
I'm just used to using switch because `is` isn't optimized:
https://github.com/dotnet/roslyn/issues/45679
We should fix that ;-)

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -445,54 +501,192 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
                     _builder.EmitOpCode(ILOpCode.Call, stackAdjustment: -3);
+                    if (used)
+                    {
+                        EmitExpression(inPlaceTarget, used: true);
```
>EmitExpression(inPlaceTarget, used: true); [](http://example.com/codeflow?start=24&length=42)

Should we be emitting constructor token before doing this?

stephentoub replied:
I can't actually get this code to execute.  Even in main, if I run all ~5000 of the code gen tests, this line:
https://github.com/dotnet/roslyn/blob/eebc8b739b34875bd3a2f6c28a3fe2c327ca6405/src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs#L2254
is never hit.  I can either leave the code as I have it, which is effectively what was there before (the line just moved into this method), or add a throw Unreachable / assert here.  What would you prefer?

AlekseyTs replied:
>>Should we be emitting constructor token before doing this?

Regardless of question how to hit this code path. What is the answer to the question? Shouldn't the call include constructor's token?

stephentoub replied:
The token is emitted below just after the call opcode:
https://github.com/dotnet/roslyn/pull/61414/files/c2797454921052f2df2fb3479f8116b4a6d88fd0#diff-1e1b32781086289e35025f3ef987f012bfe6f3fec3a25312e5f3566a977e9d98R573

AlekseyTs replied:
> the line just moved into this method

This is not how I see it, It looks to me that the line git injected in the middle of IL instruction. We haven finished emitting IL instruction for the call

stephentoub replied:
>> the line just moved into this method

> This is not how I see it

My point was that this line was never hit in main in its original location and is still never being hit.  I've deleted it in my local changes.

AlekseyTs replied:
> That's how it is today in main, too:

I do not see anything emitted between the op code and the method token in main

stephentoub replied:
Aleksey, I just said "My comments about the call and token are based on this if (used) block simply not existing, because it doesn't anymore.".

Can you please wait to continue asking questions about this until I push up my revision?

Thanks.

AlekseyTs replied:
> I just said "My comments about the call and token are based on this if (used) block simply not existing, because it doesn't anymore.".

Lets mot rush with removing the block. I'd prefer us spending some time trying to get this code path hit. The fix, I think is obvious - don't emit things in the middle of IL command, finish it and then emit the value

stephentoub replied:
> I'd prefer us spending some time trying to get this code path hit.

In main today, this line is never hit by any of the ~5000 code gen tests:
https://github.com/dotnet/roslyn/blob/eebc8b739b34875bd3a2f6c28a3fe2c327ca6405/src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs#L2254
In main, what code will cause that to be hit?

AlekseyTs replied:
We need to test this scenario anyway, whether the optimization is done or not.

stephentoub replied:
> We need to test this scenario anyway, whether the optimization is done or not.

What scenario?  I'm telling you I do not see how to hit that line.  You asked me to ensure all branches were covered, and I'm telling you I don't know how to cover an existing branch in the existing code.  Whether or not I moved it to the right exact location in this method doesn't change the fact that it's never hit by any existing test and I don't know how it's possible to get it to trigger.

So, I'm asking you for help: what code will test "this scenario"?  Forget this PR: what code will test that line in main?

stephentoub replied:
Thanks for helping me offline to come up with a way to hit it.

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -445,54 +501,192 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
                 {
-                    // consumes data ptr and size, pushes the instance
+                    // Consumes data ptr and size, pushes the instance.
                     _builder.EmitOpCode(ILOpCode.Newobj, stackAdjustment: -1);
```
>_builder.EmitOpCode(ILOpCode.Newobj, stackAdjustment: -1); [](http://example.com/codeflow?start=20&length=58)

Consider asserting `used` in this branch

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -445,54 +501,192 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+            // As we're dealing with multi-byte types, endianness needs to be considered. Such handling is provided by the
+            // runtime's RuntimeHelpers.CreateSpan, which will wrap a span around the blob on little endian and which will
+            // allocate an array and cache it on big endian.
+            MethodSymbol? createSpan = (MethodSymbol?)_module.Compilation.GetWellKnownTypeMember(WellKnownMember.System_Runtime_CompilerServices_RuntimeHelpers__CreateSpanRuntimeFieldHandle);
```
>_module.Compilation.GetWellKnownTypeMember [](http://example.com/codeflow?start=54&length=42)

Please use helper and the pattern that we use for ```rosPointerCtor```.

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -445,54 +501,192 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+            //     new ReadOnlySpan<T>(PrivateImplementationDetails.ArrayField ??= RuntimeHelpers.InitializeArray(new int[Length], PrivateImplementationDetails.DataField));
+
+            var rosArrayCtor = (MethodSymbol?)Binder.GetWellKnownTypeMember(_module.Compilation, WellKnownMember.System_ReadOnlySpan_T__ctor_Array, _diagnostics, syntax: wrappedExpression.Syntax, isOptional: true);
+            if (rosArrayCtor is null || rosArrayCtor.GetUseSiteInfo().DiagnosticInfo?.Severity == DiagnosticSeverity.Error)
```
>|| rosArrayCtor.GetUseSiteInfo().DiagnosticInfo?.Severity == DiagnosticSeverity.Erro [](http://example.com/codeflow?start=37&length=84)

Similar to ```rosPointerCtor```, this check is not necessary.

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -445,54 +501,192 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+            if (rosArrayCtor is null || rosArrayCtor.GetUseSiteInfo().DiagnosticInfo?.Severity == DiagnosticSeverity.Error)
+            {
+                // The ReadOnlySpan<T>(T[] array) constructor we need is missing or something went wrong.
+                return false;
```
>return false; [](http://example.com/codeflow?start=16&length=13)

We should try hitting this code path for existing, but bad method

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -445,54 +501,192 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+
+            ushort alignment = (ushort)specialElementType.SizeInBytes();
+            var cachingField = _builder.module.GetArrayCachingFieldForData(data, _module.Translate(arrayType), specialElementType, wrappedExpression.Syntax, _diagnostics.DiagnosticBag);
+            var arrayNotNull = new object();
```
>arrayNotNull [](http://example.com/codeflow?start=16&length=12)

arrayNotNullLabel?

AlekseyTs commented:
Done with review pass (commit 8), tests are not looked at

stephentoub commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -445,54 +501,192 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+            _builder.EmitOpCode(ILOpCode.Newobj, 0);
+            EmitSymbolToken(rosArrayCtor.AsMember(spanType), wrappedExpression.Syntax, optArgList: null);
             return true;
         }
```
Everything is exercised except:
- the `throw ExceptionUtilities.Unreachable()` when start is null != length is null
- the !_module.SupportsPrivateImplClass branch that just returns false
- some of the existing branches in https://github.com/dotnet/roslyn/blob/eebc8b739b34875bd3a2f6c28a3fe2c327ca6405/src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs#L427-L437

stephentoub commented:
All feedback addressed.

AlekseyTs commented on src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs:
```diff
@@ -53,10 +54,15 @@ internal sealed class PrivateImplementationDetails : DefaultTypeDef, Cci.INamesp
+
+        // fields for cached arrays
+        private readonly ConcurrentDictionary<(ImmutableArray<byte> Data, SpecialType ElementType), CachedArrayField> _cachedArrayFields =
+            new ConcurrentDictionary<(ImmutableArray<byte> Data, SpecialType ElementType), CachedArrayField>(CachedArrayFieldEqualityComparer.Instance);
```
>SpecialType [](http://example.com/codeflow?start=65&length=11)

I am fine with using ```SpecialType``` here, but there is an option to use ```ushort``` here and share the comparer, if you are after reducing number of types.

AlekseyTs commented on src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs:
```diff
@@ -140,38 +147,119 @@ internal void Freeze()
+            Debug.Assert(!IsFrozen);
+
+            // Get the SpecialType for the array's element type.
+            INamedTypeSymbol? symbol = arrayType.GetElementType(emitContext).GetInternalSymbol()?.GetISymbol() as INamedTypeSymbol;
```
> as INamedTypeSymbol [](http://example.com/codeflow?start=110&length=20)

When we don't expect the conversion to fail, we should use regular cast operator.

AlekseyTs commented on src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs:
```diff
@@ -140,38 +147,119 @@ internal void Freeze()
+            Debug.Assert(!IsFrozen);
+
+            // Get the SpecialType for the array's element type.
+            INamedTypeSymbol? symbol = arrayType.GetElementType(emitContext).GetInternalSymbol()?.GetISymbol() as INamedTypeSymbol;
```
>arrayType.GetElementType(emitContext).GetInternalSymbol()?.GetISymbol() [](http://example.com/codeflow?start=39&length=71)

It looks like we are mixing three different layers in this expression: Cci, internal symbol interfaces, and public symbol interfaces. We should stick with just Cci layer and not make assumptions that objects we get here implement any other layer.

stephentoub replied:
So use Cci.PrimitiveTypeCode instead of SpecialType?

AlekseyTs replied:
There is a ```ITypeReference.TypeCode``` API. There is a helper ```SpecialType SpecialTypes.GetTypeFromMetadataName(Microsoft.Cci.PrimitiveTypeCode typeCode)```, which can be used to go to ```SpecialType```. But in this layer, it is probably better to stick with PrimitiveTypeCode in order to keep API layers separate.

AlekseyTs commented on src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs:
```diff
@@ -140,38 +147,119 @@ internal void Freeze()
+                string hex = HashToHex(key.Data);
+                string name = alignment switch
+                {
+                    2 => $"{hex}2",
```
>$"{hex}2" [](http://example.com/codeflow?start=25&length=9)

What's the deal with interpolation? Can't we use a regular plus?

stephentoub replied:
We could do hex + "2". Why do you prefer it? I generally prefer interpolation because it'll generally always do the best thing now, at least as of .NET 6. In this case, it'll be lowered identically, so if there's a style preference for concatenation, fine.

AlekseyTs replied:
In general, interpolations are harder to read. One should almost become a compiler to parse it. If intent is to concatenate two string, there is nothing better than `+`.

stephentoub replied:
> interpolations are harder to read. One should almost become a compiler to parse it.

That's obviously a personal opinion and one I don't share. But I'll change it.

AlekseyTs commented on src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs:
```diff
@@ -140,38 +147,119 @@ internal void Freeze()
         }
 
-        private Cci.ITypeReference GetStorageStruct(uint size)
+        private static string HashToHex(ImmutableArray<byte> data)
```
>HashToHex [](http://example.com/codeflow?start=30&length=9)

Consider moving this function where ```GenerateDataFieldName``` used to be. That should make it easier to confirm that there is no unintentional behavior change in the code.

stephentoub replied:
I'll move it. Though fwiw is I changed its behavior a bunch of existing tests would be failing.

AlekseyTs replied:
> if I changed its behavior a bunch of existing tests would be failing

That is an optimistic assumption. Tests do not cover every possible situation. And that is one of  the reasons for having code reviews.

stephentoub replied:
Is it possible something could slip through? Of course. But especially with cryotigraphic hashes, and with a method that's just calling an existing hash function and converting the result to hex, it'd be a rare mistake that wouldn't fail every existing use of that hash.

But yes, code reviews are important.

AlekseyTs replied:
> But yes, code reviews are important.

And because of that, it helps to follow a general principle - if possible, do not move and modify code at the same time. Doing otherwise makes diff almost useless.

stephentoub replied:
I'm not going to spend more time debating this.

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -371,66 +386,135 @@ private static bool IsMultidimensionalInitializer(ImmutableArray<BoundExpression
-                if (elementType.SpecialType.SizeInBytes() != 1)
+                // This optimization is only supported for core primitive types that can be stored in metadata blobs.
+                // For enums, we need to use the underlying type.
+                specialElementType = arrayType.ElementType.EnumUnderlyingTypeOrSelf().SpecialType;
```
>arrayType.ElementType [](http://example.com/codeflow?start=37&length=21)

elementType?

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -445,54 +529,200 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+            MethodSymbol? createSpan = (MethodSymbol?)Binder.GetWellKnownTypeMember(_module.Compilation, WellKnownMember.System_Runtime_CompilerServices_RuntimeHelpers__CreateSpanRuntimeFieldHandle, _diagnostics, syntax: wrappedExpression.Syntax, isOptional: true);
+            if (createSpan is not null)
+            {
+                // CreateSpan was available. Use it.
```
>// CreateSpan was available. Use it. [](http://example.com/codeflow?start=16&length=36)

Consider adding ```Debug.Assert(!createSpan.HasUnsupportedMetadata);```. This will help to catch potential problems during testing.

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs:
```diff
@@ -445,54 +529,200 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+            // end up changing non-deterministically based on which type for a given blob was encountered first.
+            if (elementType.IsEnumType())
+            {
+                arrayType = arrayType.WithElementType(TypeWithAnnotations.Create(arrayType.ElementType.EnumUnderlyingTypeOrSelf()));
```
>arrayType.ElementType [](http://example.com/codeflow?start=81&length=21)

elementType?

AlekseyTs replied:
>TypeWithAnnotations.Create [](http://example.com/codeflow?start=54&length=26)

As we discussed offline, doing this also helps dropping custom modifiers, which potentially could be the source of similar non-deterministic behavior. I would call ```WithElementType``` unconditionally. That is unlikely to have any observable performance effect. So, trying to avoid creation of a new array symbol is probably not worth it.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadOnlySpanConstructionTest.cs:
```diff
@@ -908,5 +927,1328 @@ .maxstack  4
+        return result;
+    }}
+}}";
+            var compilation = CreateCompilationWithMscorlibAndSpan(csharp, options: TestOptions.UnsafeReleaseDll);
```
>var compilation = CreateCompilationWithMscorlibAndSpan(csharp, options: TestOptions.UnsafeReleaseDll); [](http://example.com/codeflow?start=12&length=102)

Please restore removed empty lines throughout the file and consider keeping the changes only to the necessary minimum

AlekseyTs replied:
Scratch that. I was looking at a partial diff and didn't realize this is a change to a brand ne code

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadOnlySpanConstructionTest.cs:
```diff
@@ -725,7 +817,7 @@ public void StaticFieldIsUsedForSpanCreatedFromArrayWithInitializer_02()
         {
             // This IL applies CompilerFeatureRequiredAttribute to WellKnownMember.System_ReadOnlySpan_T__ctor_Pointer.
```
>This IL applies CompilerFeatureRequiredAttribute to WellKnownMember.System_ReadOnlySpan_T__ctor_Pointer. [](http://example.com/codeflow?start=15&length=104)

Can we add similar test for System_Runtime_CompilerServices_RuntimeHelpers__CreateSpanRuntimeFieldHandle?

AlekseyTs commented:
Done with review pass (commit 10)

stephentoub commented:
I addressed all of the feedback, except for adding CompilerFeatureRequiredAttribute-based test for System_Runtime_CompilerServices_RuntimeHelpers__CreateSpanRuntimeFieldHandle. I tried following the same approach used in the other tests, but I end up with errors due to two definitions from RuntimeHelpers, one from the test and one from corelib.  Seems like I'd need to make the test not reference corelib and instead define my own version of corelib, in IL, so that the RuntimeHelpers.CreateSpan method could be attributed with CompilerFeatureRequiredAttribute, which can't be done in C#.  Have I understood correctly?  Is it really worthwhile?

AlekseyTs commented on src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs:
```diff
@@ -140,38 +147,94 @@ internal void Freeze()
-            Cci.ITypeReference type = _proxyTypes.GetOrAdd((uint)data.Length, GetStorageStruct);
-            return _mappedFields.GetOrAdd(data, data0 =>
+
+            // Get the type code for the array's element type.  It'll be used part of the
```
>It'll be used part of the [](http://example.com/codeflow?start=64&length=25)

It looks like the sentence is not finished

stephentoub replied:
I have no idea what I intended to write as that second sentence :) I'm just going to delete it.

AlekseyTs commented on src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs:
```diff
@@ -276,23 +339,31 @@ public Cci.IUnitReference GetUnit(EmitContext context)
-                c[i++] = Hexchar(b >> 4);
-                c[i++] = Hexchar(b & 0xF);
+                int i = 0;
+                foreach (var b in source.AsSpan())
```
>source.AsSpan() [](http://example.com/codeflow?start=34&length=15)

Is there a specific reason behind calling ```AsSpan``` here?

stephentoub replied:
There's less overhead to enumerating a span, as it's recognized by the runtime, bounds checks are eliminated, etc. e.g.
```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System.Collections.Immutable;
using System.Linq;

[MemoryDiagnoser]
public partial class Program
{
    static void Main(string[] args) => BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly).Run(args);

    private ImmutableArray<byte> _bytes = ImmutableArray.CreateRange(Enumerable.Range(0, 100).Select(i => (byte)i));

    [Benchmark(Baseline = true)]
    public int Sum()
    {
        int sum = 0;
        foreach (byte b in _bytes) sum += b;
        return sum;
    }

    [Benchmark]
    public int SumSpan()
    {
        int sum = 0;
        foreach (byte b in _bytes.AsSpan()) sum += b;
        return sum;
    }
}
```
on my machine:

|  Method |     Mean |
|-------- |---------:|
|     Sum | 52.85 ns |
| SumSpan | 37.12 ns |

AlekseyTs commented:
>I addressed all of the feedback, except for adding CompilerFeatureRequiredAttribute-based test for System_Runtime_CompilerServices_RuntimeHelpers__CreateSpanRuntimeFieldHandle. I tried following the same approach used in the other tests, but I end up with errors due to two definitions from RuntimeHelpers, one from the test and one from corelib. Seems like I'd need to make the test not reference corelib and instead define my own version of corelib, in IL, so that the RuntimeHelpers.CreateSpan method could be attributed with CompilerFeatureRequiredAttribute, which can't be done in C#. 

>Have I understood correctly? Is it really worthwhile?

Ok. Let's go without regression test for this scenario. Based on code review, the current behavior should be as expected.

AlekseyTs commented:
Done with review pass (commit 11)

stephentoub commented on src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs:
```diff
@@ -140,38 +147,94 @@ internal void Freeze()
-            Cci.ITypeReference type = _proxyTypes.GetOrAdd((uint)data.Length, GetStorageStruct);
-            return _mappedFields.GetOrAdd(data, data0 =>
+
+            // Get the type code for the array's element type.  It'll be used part of the
```
```suggestion
            // Get the type code for the array's element type.
```

stephentoub commented:
Feedback addressed, thanks.

AlekseyTs commented:
@jcouv Please review the latest revision

jcouv reviewed (approved):
LGTM Thanks (iteration 12)

jcouv commented:
Thanks @stephentoub :-)

stephentoub commented:
Thank you, both, for the reviews.

stephentoub commented:
I have it ready to go, we just need the new compiler to flow in :)

codemonkey85 commented:
FYI, it looks like this change doesn't work in Blazor WebAssembly. I think it's an issue with Mono? I created a sample solution to show that it works in a Console app, and not in a Blazor WASM app: https://github.com/codemonkey85/BlazorWasmIssueWithArrayInitializationWithRelevantPrimitives
